src/base/bittorrent/torrentimpl.cpp

```diff
@@ -2279,8 +2296,14 @@ bool TorrentImpl::isMoveInProgress() const
 
 void TorrentImpl::updateStatus(const lt::torrent_status &nativeStatus)
 {
-    m_pieces.clear();
-    m_nativeStatus = nativeStatus;
+    const lt::torrent_status oldStatus = std::exchange(m_nativeStatus, nativeStatus);
+
+    if (m_nativeStatus.num_pieces != oldStatus.num_pieces)
+    {
+        const QBitArray oldPieces = std::exchange(m_pieces, LT::toQBitArray(m_nativeStatus.pieces));
+        updateFilesProgress(m_pieces ^ oldPieces);
```

Comment:
> Why XOR and not simply passing `m_pieces`?

Since only newly added pieces need to be passed in `updateFilesProgress()`. `oldPieces` contains previous set and `m_pieces` the updated one.

Reply:
Anyway the code is reformatted to avoid a confusing method parameter that has additional requirements.